### PR TITLE
feat: add solana payments to solana wallets PE-5993

### PR DESCRIPTION
**Pull request by fedellen**
This PR adds support for `solana` token top ups

### CODE REVIEW

File: src/common/factory.ts [hunk 1]
```diff
@@ -48,4 +66,73 @@ export class TurboBaseFactory {
...
+    const turboSigner =
+      typeof window !== 'undefined'
+        ? new TurboWebArweaveSigner({
+            signer,
+            logger: this.logger,
+          })
+        : new TurboNodeSigner({
+            signer,
+            logger: this.logger,
+          });
```
Comment (dtfiedler) [inline src/common/factory.ts:113]: this was previously only done in the web exported version - any reason for moving it up here?

Comment (fedellen) [reply src/common/factory.ts:113]: the code in these factories seemed completely duplicated. I'll re-review that

Comment (fedellen) [reply src/common/factory.ts]: changed that back... looks like it was causing build errors 👀

File: src/node/factory.ts [hunk 2]
```diff
@@ -64,16 +71,26 @@ export class TurboFactory extends TurboBaseFactory {
             logger: this.logger,
           });
 
+    token ??= 'arweave'; // default to arweave if token is not provided
+    if (!tokenTools) {
+      tokenTools = tokenMap[token]?.({
+        gatewayUrl,
+        logger: this.logger,
+      });
+    }
```
Comment (fedellen) [inline src/node/factory.ts:80]: if token does not exist by now, being provided by user or derived from signer, then use arweave. sets the token tools if they exist for the token. if it doesn't, all othr methods that dont depend on token tools will still work

File: src/web/factory.ts [hunk 3]
```diff
@@ -46,26 +50,42 @@ export class TurboFactory extends TurboBaseFactory {
         }
       }
     } else if (privateKey) {
-      signer = new ArweaveSigner(privateKey);
+      if (token === 'solana') {
```
Comment (fedellen) [inline src/web/factory.ts:53]: this code is mostly duplicated, but the imports we're using DO matter for the build

File: src/common/token.ts [hunk 4]
```diff
@@ -190,3 +203,145 @@ export class ArweaveToken implements TokenTools {
 export const WinstonToTokenAmount = (winston: BigNumber.Value) => winston;
 export const ARToTokenAmount = (ar: BigNumber.Value) =>
   new BigNumber(ar).times(1e12).valueOf();
+
+export class SolanaToken implements TokenTools {
```
Comment (dtfiedler) [inline src/common/token.ts:207]: nit: put these in separate files `tokens/solana.ts`

File: src/common/token.ts [hunk 5]
```diff
@@ -190,3 +203,145 @@ export class ArweaveToken implements TokenTools {
...
+
+    await this.connection.sendRawTransaction(tx.serialize(), {
+      maxRetries: this.pollingOptions.maxAttempts,
+    });
+
+    await this.connection.confirmTransaction(
+      {
+        signature: id,
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        blockhash: tx.recentBlockhash!,
```
Comment (dtfiedler) [inline src/common/token.ts:273]: can we just do a check above if it does not exist to avoid this null assertion? same as below?

**Comment by dtfiedler** [comment]
my primary concerns here are the ways in which we're using signers for their respective signers - ideally we have an abstract class in common and the node/web implementations properly import what they need for their given environments.

File: src/web/factory.ts [hunk 6]
```diff
@@ -31,13 +32,16 @@ export class TurboFactory extends TurboBaseFactory {
     signer: providedSigner,
     paymentServiceConfig = {},
     uploadServiceConfig = {},
-    tokenMap,
```
Comment (fedellen) [inline src/web/factory.ts:34]: from huddle: keep this here but deprecate it for a v2

File: src/node/factory.ts [hunk 7]
```diff
@@ -47,7 +49,12 @@ export class TurboFactory extends TurboBaseFactory {
         }
       }
     } else if (privateKey) {
-      signer = new ArweaveSigner(privateKey);
+      if (token === 'solana') {
+        // TODO: test wallet creation
+        signer = new HexSolanaSigner(privateKey);
+      } else {
+        signer = new ArweaveSigner(privateKey);
+      }
```
Comment (fedellen) [inline src/node/factory.ts:57]: from huddle:

- move the below to `/common/factory`
- implement separate function for getting signer

File: src/types.ts [hunk 8]
```diff
@@ -50,13 +50,9 @@ export type Country = 'United States' | 'United Kingdom' | 'Canada'; // TODO: ad
...
 export type CreditableTokenType = (typeof allowedFiatTokens)[number];
 
-export const tokenTypes = ['arweave' /*'solana', 'ethereum'*/] as const;
+export const tokenTypes = ['arweave', 'solana' /* 'ethereum'*/] as const;
 export type TokenType = (typeof tokenTypes)[number];
 
-export type TokenMap = {
-  [key in TokenType]: TokenTools;
-};
-
```
Comment (fedellen) [inline src/types.ts:59]: Bring type back as @deprecated

File: tests/turbo.web.test.ts [hunk 9]
```diff
@@ -562,10 +560,20 @@ describe('Browser environment', () => {
 
     const signer = new HexSolanaSigner(testSolWallet);
 
+    const tokenTools = new SolanaToken({
+      gatewayUrl: 'https://api.devnet.solana.com',
```
Comment (fedellen) [inline tests/turbo.web.test.ts:564]: add a variable in helpers.ts

File: src/common/token.ts (hunk 5 above)
Comment (fedellen) [reply src/common/token.ts]: 929bd19

File: src/common/token.ts (hunk 4 above)
Comment (fedellen) [reply src/common/token.ts]: 3b9827f

File: src/types.ts (hunk 8 above)
Comment (fedellen) [reply src/types.ts:59]: 7d651b8

File: src/web/factory.ts (hunk 6 above)
Comment (fedellen) [reply src/web/factory.ts:34]: 7d651b8

File: tests/turbo.web.test.ts (hunk 9 above)
Comment (fedellen) [reply tests/turbo.web.test.ts]: 696a5f0

File: src/node/factory.ts (hunk 7 above)
Comment (fedellen) [reply src/node/factory.ts]: [10d0219](https://github.com/ardriveapp/turbo-sdk/pull/113/commits/10d0219dd1d3a038a04e8669c0564e7b3f4b7967)

File: src/node/factory.ts [hunk 10]
```diff
@@ -14,70 +14,6 @@
...
-      signer: turboSigner,
-      logger: this.logger,
-    });
-    return new TurboAuthenticatedClient({
-      uploadService,
-      paymentService,
-    });
-  }
-}
+export class TurboFactory extends TurboBaseFactory {}
```
Comment (fedellen) [inline src/node/factory.ts:19]: logic here moved to `/common/factory`

File: src/common/token/solana.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+      try {
+        status = await this.connection.getSignatureStatus(txId);
+      } catch (err) {
+        // Continue retries when request errors
+        this.logger.debug('Failed to poll for transaction...', { err });
+      }
+
+      if (status && status.value && status.value.err !== null) {
+        throw new Error(`Transaction failed: ${status.value.err}`);
```
Comment (fedellen) [inline src/common/token/solana.ts:152]: test coverage: could stub an error'ed SOL Tx here

File: src/common/token/solana.ts [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    });
+
+    if (
+      tx.recentBlockhash === undefined ||
+      tx.lastValidBlockHeight === undefined
+    ) {
+      throw new Error(
+        'Failed to submit Transaction --  missing blockhash or lastValidBlockHeight from transaction creation. Solana Gateway Url:' +
+          this.gatewayUrl,
+      );
```
Comment (fedellen) [inline src/common/token/solana.ts:111]: test coverage: could stub getLatestBlockhash to retrun undefined

File: src/common/token/arweave.ts [hunk 13]
```diff
@@ -164,14 +167,15 @@ export class ArweaveToken implements TokenTools {
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public async submitTx(tx: any): Promise<void> {
     try {
       const response = await this.arweave.transactions.post(tx);
 
       if (response.status !== 200) {
         throw new Error(
           'Failed to post transaction -- ' +
-            `Status ${response.status}, ${response.statusText}`,
+            `Status ${response.status}, ${response.statusText}, ${response.data}`,
```
Comment (fedellen) [inline src/common/token/arweave.ts:178]: test coverage: could stub AR tx to return wrong status

File: src/common/factory.ts [hunk 14]
```diff
@@ -48,4 +70,94 @@ export class TurboBaseFactory {
...
+        token = 'arweave';
+      }
+    }
+
+    const turboSigner = this.getSigner(providedSigner, privateKey, token);
+
+    if (!tokenTools) {
+      if (tokenMap && token === 'arweave') {
+        tokenTools = tokenMap.arweave;
+      }
```
Comment (fedellen) [inline src/common/factory.ts:138]: test coverage: we can add backwards compatibility tests for this case

File: src/common/factory.ts [hunk 15]
```diff
@@ -48,4 +70,94 @@ export class TurboBaseFactory {
...
+    token: CreditableTokenType,
+  ): TurboDataItemAbstractSigner {
+    let signer: TurboSigner;
+
+    if (providedSigner !== undefined) {
+      signer = providedSigner;
+    } else if (providedPrivateKey !== undefined) {
+      if (token === 'solana') {
+        signer = new HexSolanaSigner(providedPrivateKey);
+
```
Comment (fedellen) [inline src/common/factory.ts:86]: test coverage: we can cover this with providing secret key to factory

File: src/common/factory.ts [hunk 16]
```diff
@@ -48,4 +70,94 @@ export class TurboBaseFactory {
...
+    } else {
+      throw new Error('A privateKey or signer must be provided.');
+    }
+
+    if (typeof window !== 'undefined') {
+      return new TurboWebArweaveSigner({
+        signer,
+        logger: this.logger,
+      });
+    }
```
Comment (fedellen) [inline src/common/factory.ts:101]: test coverage: Could force `window` in node env to cause this

**Comment by dtfiedler** [comment]
if you do `yarn link && yarn build` and then run `yarn example:esm` and `yarn example:cjs` do they still work as expected?

**Comment by fedellen** [comment]
For esm, I get

```sh

                  ^

Error: Failed to post transaction -- Failed to post transaction -- Status 400, Transaction verification failed.
    at ArweaveToken.submitTx (file:///Users/derek/ar-io/ardrive-turbo-sdk/examples/esm/node_modules/@ardrive/turbo-sdk/lib/esm/common/token.js:118:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async TurboAuthenticatedPaymentService.topUpWithTokens (file:///Users/derek/ar-io/ardrive-turbo-sdk/examples/esm/node_modules/@ardrive/turbo-sdk/lib/esm/common/payment.js:190:9)
    at async file:///Users/derek/ar-io/ardrive-turbo-sdk/examples/esm/index.mjs:87:23
    ```
    
    which would be expected from a wallet with no balance
    
    but cjs I'm getting an import error on this helper
    
    ```sh
                     ^

TypeError: WinstonToTokenAmount is not a function
    at /Users/derek/ar-io/ardrive-turbo-sdk/examples/cjs/index.cjs:91:18
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by fedellen** [comment]
but wouldn't yarn build not effect the outcome? that packkage json pulls from `latest`

**Comment by fedellen** [comment]
hmmm. fwiw when I build loacally the import resolves in CJS. 
 
 ```sh
 
/Users/derek/ar-io/ardrive-turbo-sdk/examples/cjs/node_modules/@ardrive/turbo-sdk/lib/cjs/common/token/arweave.js:127
            throw new Error(`Failed to post transaction -- ${err instanceof Error ? err.message : err}`);
                  ^

Error: Failed to post transaction -- Failed to post transaction -- Status 400, Transaction verification failed., [object Object]
    at ArweaveToken.submitTx (/Users/derek/ar-io/ardrive-turbo-sdk/examples/cjs/node_modules/@ardrive/turbo-sdk/lib/cjs/common/token/arweave.js:127:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
 ```
 
 not sure I understand, the imports look okay

**Comment by dtfiedler** [comment]
> but wouldn't yarn build not effect the outcome? that packkage json pulls from `latest`

you're right, you'd need to go into each and run `yarn link @ardriveapp/turbo-sdk` so they use the local build

**Comment by fedellen** [comment]
using `../..` in the package json seems to work

**Review by dtfiedler — APPROVED** [review]

**Comment by dtfiedler** [comment]
let's heavily test this in alpha against our examples to make sure we didn't unintentionally break any signing behavior

**Pull request merged**